ml-agents/mlagents/trainers/optimizer/torch_optimizer.py

```diff
@@ -49,25 +55,115 @@ def create_reward_signals(self, reward_signal_configs):
                 reward_signal, self.policy.behavior_spec, settings
             )
 
+    def _evaluate_by_sequence(
+        self, tensor_obs: List[torch.Tensor], initial_memory: np.ndarray
+    ) -> Tuple[Dict[str, torch.Tensor], AgentBufferField, torch.Tensor]:
+        """
+        Evaluate the batch sequence-by-sequence, assembling the result. This enables us to get the
+        intermediate memories for the critic.
+        """
+        num_experiences = tensor_obs[0].shape[0]
+        all_next_memories = AgentBufferField()
+        # In the buffer, the 1st sequence are the ones that are padded. So if seq_len = 3 and
+        # trajectory is of length 10, the 1st sequence is [pad,pad,obs].
+        # Compute the number of elements in this padded seq.
+        leftover = num_experiences % self.policy.sequence_length
+
+        # Compute values for the potentially truncated initial sequence
+        _mem = initial_memory
+        seq_obs = []
+
+        first_seq_len = self.policy.sequence_length
+        for _obs in tensor_obs:
+            if leftover > 0:
+                first_seq_obs = _obs[0:leftover]
+                first_seq_len = leftover
+            else:
+                first_seq_obs = _obs[0 : self.policy.sequence_length]
+            seq_obs.append(first_seq_obs)
+
+        for _ in range(first_seq_len):
+            all_next_memories.append(initial_memory.squeeze().detach().numpy())
```

Comment:
```suggestion
            all_next_memories.append(_mem.squeeze().detach().numpy())
```

Reply:
Hard to understand why the all_next_memories are a concatenation of initial memories...

Reply:
Refactored this to be a bit clearer